llvm/lib/Support/Windows/Path.inc

```diff
@@ -1387,12 +1387,33 @@ std::error_code remove_directories(const Twine &path, bool IgnoreErrors) {
   Path16.push_back(0);
   Path16.push_back(0);
 
-  SHFILEOPSTRUCTW shfos = {};
-  shfos.wFunc = FO_DELETE;
-  shfos.pFrom = Path16.data();
-  shfos.fFlags = FOF_NO_UI;
+  HRESULT HR =
+      CoInitializeEx(NULL, COINIT_APARTMENTTHREADED | COINIT_DISABLE_OLE1DDE);
+  if (SUCCEEDED(HR)) {
+    IFileOperation *FileOp = NULL;
+    HR = CoCreateInstance(CLSID_FileOperation, NULL, CLSCTX_ALL,
+                          IID_PPV_ARGS(&FileOp));
+    if (SUCCEEDED(HR)) {
+      HR = FileOp->SetOperationFlags(FOF_NO_UI | FOFX_NOCOPYHOOKS);
+      if (SUCCEEDED(HR)) {
+        PIDLIST_ABSOLUTE PIDL = ILCreateFromPathW(Path16.data());
+        IShellItem *ShItem = NULL;
+        HR = SHCreateItemFromIDList(PIDL, IID_PPV_ARGS(&ShItem));
+        if (SUCCEEDED(HR)) {
+          HR = FileOp->DeleteItem(ShItem, NULL);
+          if (SUCCEEDED(HR)) {
+            HR = FileOp->PerformOperations();
```

Comment:
We generally try to avoid these kinds of deeply nested blocks. It looks like this could be make mostly linear with the use of [early returns](https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code) (possibly together with some helper functions) and `make_scope_exit`  to run cleanups.

Reply:
Yes, I know. It is possible to use CComPtr<> for IFileOperation and IShellItem, but not for PIDLIST_ABSOLUTE. Otherwise it is necessary to add a lot of release code before each early return or some complicated cleanup code at the end with few checks. This implementation is the most compact.

Reply:
It may be compact but it goes directly against the coding standards. And you don't need to add releases before every early returns -- that's what make_scope_exit is for. I don't know if it's possible to use some COM apis for that or not, but something like this should be equivalent to your code, but have a nesting level of one.

```
HRESULT deleteDirectoriesImpl(...) {
  if (HRESULT HR =
      CoInitializeEx(NULL, COINIT_APARTMENTTHREADED | COINIT_DISABLE_OLE1DDE); !SUCCEEDED(HR))
    return HR;
  auto Uninitialize = make_scope_exit([]{CoUninitialize();});
  IFileOperation *FileOp = NULL;
  if (HRESULT HR = CoCreateInstance(CLSID_FileOperation, NULL, CLSCTX_ALL,
                          IID_PPV_ARGS(&FileOp)); !SUCCEEDED(HR))
    return HR;
  auto ReleaseFile = make_scope_exit([]{FileOp->Release();});
  ...
```

Reply:
How to ensure the release order w/o scope {}?
```
FileOp->Release();
CoUninitialize();
```
[Use Early Exits and continue to Simplify Code](https://llvm.org/docs/CodingStandards.html#id40) means use it when it makes sense, does not complicate the code, and does not make it less stable and predictable.

Reply:
c++ always runs destructors in the [opposite construction order](https://isocpp.org/wiki/faq/dtors#order-dtors-for-locals). There's nothing unstable or unpredictible about that. I'm sure you can make a case that the the code you wrote is simpler, but I don't think most devs would agree with you.

Reply:
Updated.

Reply:
Thanks, I like the new code.